[Detached] Make node-geolite2 a redistribution of MaxMind's databases

GitSquared opened the pull request:
Following the discussion in #17.

Roadmap before mergeable:
 - [x] Track databases on GitHub
 - [x] Rewrite postinstall script
 - [x] Write new lib License
 - [x] Setup automatic updates, "backend-side" ([see this comment](https://github.com/runk/node-geolite2/issues/17#issuecomment-570535318)) either with Travis or GitHub Actions, and a maintainer's MaxMind license
 - [X] Setup automatic updates, "client-side" - make the library keep the downloaded databases up to date (I added checksum validation which could be useful for this)
 - [X] Send an email to [MaxMind's legal](https://github.com/runk/node-geolite2/issues/17#issuecomment-570569660) to make sure they're OK with what we're doing (should be, but won't hurt to ask)
 - [x] Update README
 - [x] Write new tests
 - [X] ~~Change the [download URL](https://github.com/runk/node-geolite2/blob/3e520a90d83d3966fa88683aede389d5b13fd238/scripts/download-helper.js#L8-L9) to point to this repo instead of my fork~~ Detach fork and publish under new name to NPM

I'll keep working on this but I wanted to share my progress so far and get @runk's suggestions if any.

---

Close #17

AnandChowdhary commented on package.json:
```diff
@@ -28,6 +28,7 @@
   },
   "homepage": "https://github.com/runk/node-geolite2#readme",
   "dependencies": {
+    "md5-file": "^4.0.0",
```
Maybe it's a good idea to use a SHA-2 hash like SHA-384 instead of MD5? It's more future-proof and is recommended for resource Integrity by W3C, as in this spec: https://w3c.github.io/webappsec-subresource-integrity/#resource-integrity.

GitSquared replied:
Sure. I used md5 because that's what MaxMind use for their checksums but didn't think too much about it. I'l' switch to SHA-384

GitSquared replied:
62e752f

GitSquared commented:
So I implemented a self-updating mechanism to automatically update the local databases copies from the git repo. It uses SHA-384 checksums to determine whether a newer version is available. It's capable of keeping database copies up to date transparently, even when the library is used in a long-running script like a server or something, which should help ensure that nobody keeps using old versions of the databases past the allowed upgrade period of 30 days (see Do Not Sell requests and relevant sections in MaxMind's EULA).

I had to make a major API change however: the [`maxmind`](https://github.com/runk/node-maxmind) module is now incorporated in `geolite2`.

Old API:
```js
var geolite2 = require('geolite2');
var maxmind = require('maxmind');

var lookup = maxmind.openSync(geolite2.paths.city); // or geolite2.paths.country or geolite2.paths.asn
var city = lookup.get('66.6.44.4');
```

Proposed API (this PR):
```js
var geolite2 = require('geolite2');

var lookup = geolite2.openSync('GeoLite2-City'); // or -Country or -ASN
// an asynchronous geolite2.open() method is also available and returns a Promise
var city = lookup.get('66.6.44.4');
```

<details>
<summary>Implementation details</summary>
geolite2 returns a Proxy wrapper around a maxmind Reader instance. When a property of the reader is accessed, a background routine starts (either once every 48 hours or the first time the reader is accessed) to fetch the latest databases checksums from the git mirror.

If any of the checksums differ from the local copy, the databases are fetched again, and their checksums matched again. Normally they should now be correct, so new maxmind Reader instances are spun up and replace the old ones.

If the checksums don't match after a fresh download, or for any reason anything fucks up, a warning is printed to stdout and the whole self-update shebang will run again with the usual trigger (48 hours or program restart).
</details>

CI is broken again because the tests are now irrelevant and need to be rewritten.

---

> Setup automatic updates, "backend-side" ([see this comment](https://github.com/runk/node-geolite2/issues/17#issuecomment-570535318)) either with Travis or GitHub Actions, and a maintainer's MaxMind license

@runk I'm going to need your guidance/help for this part; as this involves maintainer secrets and CI configuration that I cannot change within the scope of a PR.

Once this is done we just have to verify the redistribution terms with MaxMind and, assuming all parties involved are OK, this will be mergeable.

runk commented on index.js:
```diff
@@ -1,9 +1,61 @@
-var path = require("path");
+const fs = require('fs');
+const path = require('path');
+const maxmind = require('maxmind');
```
Hmm, I don't mind it, however feels that consumers of this library should have a freedom of choice which lookup module they use. Thoughts?

AnandChowdhary replied:
[maxmind](https://www.npmjs.com/package/maxmind) is another @runk package and not an official package from MaxMind (the company), though the name might seem that way. This is why I agree — we should let people use other packages that can search `.mmmd` files with this library (whose job is then to just prepare the databases but not query them).

GitSquared replied:
Fair point. Perhaps the `geolite2.open` and `geolite2.openSync` methods could request a second argument, that would be a function to build a database reader from a database path?

Example (end-user code):

```js
(async () => {
	const geolite2 = require('geolite2');
	const maxmind = require('maxmind'); // Or any other reader lib
	const fs = require('fs');
	
	// Second argument is a function to build a db reader from the file path
	// geolite2 will automatically rebuild the reader with the provided function on database updates
	const lookup = await geolite2.open('GeoLite2-City', async path => {
		return await maxmind.open(path); // Or custom lib code
		// (the returned object will be wrapped by geolite2's self-updating Proxy, should work with any object)
	});

	// Synchronous version
	const syncLookup = geolite2.openSync('GeoLite2-City', path => {
		return new maxmind.Reader(fs.readFileSync(path));
	});

	console.log(lookup.get('8.8.8.8'));
	console.log(syncLookup.get('8.8.8.8'));
})();
```

What do you think of this?

runk replied:
I'd personally leave this detail for consumers. You might emit a signal that database file has changed, and they'd have their own means of ensuring they refreshed an instance of their lookup client or whatever.

GitSquared replied:
I have thought of this too - basically only providing an EventEmitter.
However since refreshing the reader instance on found updates is required by the license, I figured most users would end up with something looking like this code so we might as well do it in the lib.

Would it be OK with you if we allowed both the API I described above, and an EventEmitter-based one for more in-depth performance optimization/database usage in apps that need it?

GitSquared replied:
I looked into this again and realized adding events didn't make much sense. I'm going to push an updated API & Readme, feel free to open a new comment if you dislike that solution.

GitSquared replied:
fafb2b0

GitSquared replied:
8b9fdcc Added the EventEmitter API, documented.

runk commented on scripts/download-helper.js:
```diff
@@ -0,0 +1,101 @@
+  { name: 'GeoLite2-ASN' },
+  { name: 'GeoLite2-City' },
+  { name: 'GeoLite2-Country' }
+];
```
Perhaps i'd do `const editions = [ ... ].map((edition) => ...)`

GitSquared replied:
Indeed, that's much better. I changed the code to use Array.map() whenever possible.

runk commented on scripts/download-helper.js:
```diff
@@ -0,0 +1,101 @@
+});
+
+function fetchChecksums() {
+  let downloads = [];
```
`downloads = editions.map(...)`

runk commented on scripts/download-helper.js:
```diff
@@ -0,0 +1,101 @@
+        res.on('end', () => {
+          checksum = checksum.trim();
+          if (!res.complete || checksum.length !== 96) throw new Error(`Could not fetch checksum for ${edition.name}\n\nReceived:\n${checksum}`);
+          edition.checksum = checksum;
```
I'm not a big fan of monkey-patching in real time. Any way we can avoid it?

GitSquared replied:
I'm not sure I understand your comment here - what monkey-patching are you referring to?

runk replied:
You're extending items in `editions` list asynchronously

GitSquared replied:
Got it (I think). Will fix this to only change `editions` if all checksums have successfully been retrieved.

GitSquared replied:
7d2263c

runk commented on scripts/download-helper.js:
```diff
@@ -0,0 +1,101 @@
+  let downloads = [];
+
+  editions.forEach(edition => {
+    downloads.push(new Promise(resolve => {
```
resolve is not being called?

runk commented on scripts/download-helper.js:
```diff
@@ -0,0 +1,101 @@
+  editions.forEach(edition => {
+    downloads.push(new Promise(resolve => {
+      fetch(edition.dbURL).then(res => {
+        res.pipe(tar.t()).on('entry', entry => {
```
perhaps it'd be nice to have un-tarring as part of the iterator (i.e. combine with `fetch`)

runk commented on scripts/download-helper.js:
```diff
@@ -0,0 +1,101 @@
+    promises.push(new Promise((resolve, reject) => {
+      fs.readFile(path.join(downloadPath, edition.name+'.mmdb'), (err, buffer) => {
+        if (err) reject(err);
+        hash = sha384(buffer).toString('hex');
```
I cannot see where variable is defined; Perhaps do it inline like `if (sha384(buffer).toString('hex') === edition.checksum)`?

runk commented on scripts/download-helper.js:
```diff
@@ -0,0 +1,101 @@
+        if (hash === edition.checksum) {
+          resolve();
+        } else {
+          reject(`Mismatched checksums for ${edition.name}`);
```
reject(new Error( ... ))

AnandChowdhary replied:
Reference for @runk's comment: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject#Examples

GitSquared replied:
:+1:

GitSquared commented:
Thanks for taking the time to review this - lots of new code indeed.
Regarding the CI/backend stuff, I'm assuming we'd rather go with Travis since it's already set up - is that okay?

I'll write a bash script to be run as a cron job. Need to look into how to push from CI.

GitSquared commented:
Update script written. Ci-independent for now, I'll add that part later on.

```
❯ env MAXMIND_LICENSE_KEY=[key] ./update-databases

Downloading checksum for latest version of GeoLite2-ASN...
Downloading GeoLite2-ASN...
Verifying checksum...
Extracting database...
Calculating database checksum for redistribution...
Database GeoLite2-ASN successfully updated

Downloading checksum for latest version of GeoLite2-Country...
Downloading GeoLite2-Country...
Verifying checksum...
Extracting database...
Calculating database checksum for redistribution...
Database GeoLite2-Country successfully updated

Downloading checksum for latest version of GeoLite2-City...
Downloading GeoLite2-City...
Verifying checksum...
Extracting database...
Calculating database checksum for redistribution...
Database GeoLite2-City successfully updated

All databases processed
```

runk commented:
FYI there's `MAXMIND_LICENSE_KEY` env var already setup on travis ci. Probably you already figured it :)

GitSquared commented:
Indeed, which is why I reused that env var in the bash script.
By the way, it checks current versions against the remote checksums before pulling updates, to ensure your license key isn't over-used and flagged as spam or whatever.

In other news, I have written a draft mail for MM's legal, if you're OK with that I'd like to send it to you first so you can review it.

runk commented:
Sure 👍 My email is on the profile

GitSquared commented:
The Travis cron job script is up.
@runk Once this merges, you'll need to setup a [cron job](https://docs.travis-ci.com/user/cron-jobs/) on master and a `$GITHUB_PUSH_TOKEN` secret env var on Travis with a [GitHub personal access token](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line#using-a-token-on-the-command-line).

GitSquared commented:
Readme has been updated.

GitSquared commented:
Readme documentation and tests have been written.

d1str0 commented:
I'm the maintainer of an open source project that also used to directly download these databases on install (directly from Maxmind, not via this repo). I would love to hear what the response from the legal team is as this is something my project will also have to deal with.

Cheers

GitSquared commented:
@d1str0 We have not contacted them yet. Runk and I are working in opposite timezones so things are going a bit slow.
You can subscribe to this thread for further updates.

GitSquared commented:
@runk Have you received my email?

runk commented:
Hey @GitSquared, it was in trash - just found it.

After reading and looking at all the code I still feel a little uncomfortable redistributing databases like that. I really appreciate all the work you've done, however I don't think I'm feeling like accepting this change in the current repository. Whole legal deal makes me feel too uncomfortable.

That said, would you mind cloning/copying this repo and doing exactly what you want? I'm more than happy to add a reference to your repo from this one - no questions at all.

GitSquared commented:
> it was in trash - just found it.

GMail hates personal mailservers. I'd appreciate if you could report it as "not spam". Thanks.

I understand your decision, thanks for reviewing this PR either way.
I will contact MaxMind myself and detach my fork is they give me a positive answer - I'll update this thread.

I know you're not a stranger to them - they reuse your reader lib in their official APIs - so I had wished that with you as a maintainer they would be more likely to try and find a way for the license to work. Guess it's time to test my street cred.

GitSquared commented:
I got an answer from MaxMind.

They provided a helpful link to [a page deep inside their support Q&A](https://support.maxmind.com/geolite-faq/general/how-do-i-use-the-geolite2-databases-in-a-product-or-service/) along with the following statement:

> If you provide attribution you can redistribute the GeoLite2 Databases without a commercial redistribution license.

> If GeoLite2 will be distributed through your product, you need to ensure that you have communicated or provided a mechanism for your customers to use the most current version of the database, that databases are required to be destroyed 30 days after MaxMind releases a new version, and that do not sell notifications flow down to those who you’ve shared the database with.

They kept themselves in the clear by adding that they "are not in a position to give legal advice" - as far as I'm involved this is just lawyer speak to avoid any troubles.

So **tl;dr** they cool with it, especially since it self-updates.

I'll detach & publish my fork in the next few days.

d1str0 commented:
@GitSquared even though I don't use Node or these libs in my project, I highly appreciate all the work you've put into this issue, both from a coding implementation and also handling communication with Maxmind. Kudos, mate.

GitSquared commented:
@d1stro Cheers. Sometimes I like to get shit done.

GitSquared commented:
Package is live as `geolite2-redist`: [npm](https://www.npmjs.com/package/geolite2-redist) | [git](https://github.com/GitSquared/node-geolite2-redist)

@runk feel free to mention it in this one's README. You can also prob close #17

AnandChowdhary commented:
Thanks for all the hard work, @GitSquared and @runk!
